File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        
        def canShip(weight):
            current_weight = 0
            no_of_days = 1

            for amount in weights:
                current_weight += amount
                if (current_weight > weight):
                    no_of_days += 1
                    current_weight = amount
            
            return no_of_days <= days

        
        i = max(weights)
        j = sum(weights)

        while (i <= j):
            mid = (i + j) // 2

            if (canShip(mid)):
                j = mid - 1
            else:
                i = mid + 1

        
        return i

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.js
var shipWithinDays = function(weights, days) {
    let i = Math.max(...weights)
    let j = weights.reduce((a, b)=> a + b, 0)
    let min_weight = Infinity

    function canShip(capacity) {
        let currWeight = 0
        let no_of_days = 1

        for (let weight of weights) {
            currWeight += weight
            if (currWeight > capacity) {
                currWeight = weight
                no_of_days += 1
            }
        }
        return no_of_days <= days
    }    

    while (i <= j) {
        let capacity = Math.floor((i + j) / 2)
        if (canShip(capacity)) j = capacity - 1  
        else i = capacity + 1
    }


    return i
};